File: 2022-2023/sourcecode/lab/lab7/NotyJava/app/src/main/java/pl/udu/uwr/pum/notyjava/widget/NotyWidgetItemFactory.java
package pl.udu.uwr.pum.notyjava.widget;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import java.util.ArrayList;

import pl.udu.uwr.pum.notyjava.R;
import pl.udu.uwr.pum.notyjava.db.DBHandler;
import pl.udu.uwr.pum.notyjava.model.NoteModel;

class NotyWidgetItemFactory implements RemoteViewsService.RemoteViewsFactory {

    private final Context context;
    private final int appWidgetId;
    private ArrayList<NoteModel> noteList;
    private DBHandler dbHandler;

    public NotyWidgetItemFactory(Context context, Intent intent) {
        this.context = context;
        this.appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID,
                AppWidgetManager.INVALID_APPWIDGET_ID);
    }

    @Override
    public void onCreate() {
        dbHandler = new DBHandler(context);
        noteList = dbHandler.getNotes();
    }

    @Override
    public void onDataSetChanged() {
        noteList = dbHandler.getNotes();
    }

    @Override
    public void onDestroy() {
        dbHandler.close();
    }

    @Override
    public int getCount() {
        return noteList.size();
    }

    @Override
    public RemoteViews getViewAt(int position) {
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.item_list);
        remoteViews.setTextViewText(R.id.itemListTextView, noteList.get(position).getTime().toString()
                + "\n" + noteList.get(position).getTextNote());
        remoteViews.setTextColor(R.id.itemListTextView, noteList.get(position).getColor());

        Intent fillIntent = new Intent();
        fillIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        fillIntent.putExtra("id", noteList.get(position).getId());
        remoteViews.setOnClickFillInIntent(R.id.itemListTextView, fillIntent);
        return remoteViews;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int position) {
        return noteList.get(position).getId();
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }
}

File: 2022-2023/sourcecode/lab/lab7/NotyKotlin/app/src/main/java/pl/udu/uwr/pum/notykotlin/widget/NotyWidgetItemFactory.kt
package pl.udu.uwr.pum.notykotlin.widget

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import pl.udu.uwr.pum.notykotlin.R
import pl.udu.uwr.pum.notykotlin.db.DBHandler
import pl.udu.uwr.pum.notykotlin.model.NoteModel

class NotyWidgetItemFactory(private val context: Context, intent: Intent) :
    RemoteViewsService.RemoteViewsFactory {
    private lateinit var dbHandler: DBHandler
    private val appWidgetId: Int
    private lateinit var noteList: List<NoteModel>

    override fun onCreate() {
        dbHandler = DBHandler(context)
        noteList = dbHandler.notes
    }

    override fun onDataSetChanged() {
        noteList = dbHandler.notes
    }

    override fun onDestroy() {
        dbHandler.close()
    }

    override fun getCount(): Int = noteList.size

    override fun getViewAt(position: Int): RemoteViews {
        val remoteViews = RemoteViews(context.packageName, R.layout.item_list)
        remoteViews.setTextViewText(
            R.id.itemListTextView, noteList[position].time.toString() + "\n" + noteList[position].textNote
        )
        remoteViews.setTextColor(R.id.itemListTextView, noteList[position].color)

        val fillIntent = Intent()
        fillIntent.apply {
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
            putExtra("id", noteList[position].id)
        }
        remoteViews.setOnClickFillInIntent(R.id.itemListTextView, fillIntent)
        return remoteViews
    }

    override fun getLoadingView(): RemoteViews? {
        return null
    }

    override fun getViewTypeCount(): Int = 1

    override fun getItemId(position: Int): Long = noteList[position].id.toLong()

    override fun hasStableIds(): Boolean = true

    init {
        appWidgetId = intent.getIntExtra(
            AppWidgetManager.EXTRA_APPWIDGET_ID,
            AppWidgetManager.INVALID_APPWIDGET_ID
        )
    }
}